### PR TITLE
Add docs about prod ecosystem features

### PR DESCRIPTION
**Pull request by dzhulgakov**
Covering fleet-wide profiling, api logging, etc.

It's my first time writing rst, so suggestions are definitely welcomed.

### CODE REVIEW

File: docs/source/notes/production_ecosystem.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
+Integration in production ecosystem
```
[the remark this answers is not in the document]
Comment (soumith) [reply docs/source/notes/production_ecosystem.rst]: Integration in production environment
-> Features for large-scale deployments of PyTorch

File: docs/source/notes/production_ecosystem.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+===================================
+
```
Comment (soumith) [inline docs/source/notes/production_ecosystem.rst]: Add a table of contents here

File: docs/source/notes/production_ecosystem.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+.. code-block:: cpp
+
+    SetExportModuleExtraFilesHook([](const script::Module&) {
+        script::ExtraFilesMap files;
+        files["producer_info.json"] = "{\"user\": \"" + getenv("USER") + "\"}";
+        return files;
+    });
+
+
+API usage logging
```
Comment (soumith) [inline docs/source/notes/production_ecosystem.rst]: move this section above "Attaching metadata to saved TorchScript models" as it flows much better after the operator profiling section

**Review by soumith — CHANGES_REQUESTED** [review]

File: docs/source/notes/production_ecosystem.rst (hunk 2 above)
Comment (dzhulgakov) [reply docs/source/notes/production_ecosystem.rst]: There's already a table of contents rendered on the right, but sure

File: docs/source/notes/large_scale_deployments.rst [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            &onFunctionEnter,
+            &onFunctionExit,
+            /* needs_inputs */ true,
+            /* sampled */ true
+        );
+    }
+
+    void onFunctionEnter(const RecordFunction& fn) {
+        std::cerr << "Before function " << fn.name() 
+                  << " with " << fn.inputs().size() << "inputs" << std::endl;
```
Comment (ilia-cher) [inline docs/source/notes/large_scale_deployments.rst]: nit: space before inputs

File: docs/source/notes/large_scale_deployments.rst [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    void onFunctionExit(const RecordFunction& fn) {
+        std::cerr << "After function " << fn.name();
+    }
+
+API usage logging
+^^^^^^^^^^^^^^^^^
+
+When running in a broader ecosystem, for example in managed job scheduler, it's
+often useful to track which binaries invoke particular PyTorch APIs. There
+exists simple intrumentation injected at several important API points that
```
Comment (ilia-cher) [inline docs/source/notes/large_scale_deployments.rst]: typo: instrumentation

**Review by ilia-cher — COMMENTED** [review]
Checked the part on operator callbacks - LGTM, please also add that pushCallback and setSamplingProbability are not thread safe and should be called during initialization only and should never be called while any operator is running.

**Review by ilia-cher — APPROVED** [review]

File: docs/source/notes/large_scale_deployments.rst [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+Attaching metadata to saved TorchScript models
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+TorchScript modules can be saved as an archive file that bundles serialized
+parameters and module code as TorchScript (see :meth:`torch.jit.save`). It's
+often convenient to bundle additional information together with the model, for
+example, description of model producer or auxiliary artifacts.
+
+It can be achieved by passing the ``_extra_files`` argument to
```
Comment (apaszke) [inline docs/source/notes/large_scale_deployments.rst]: Well we should never document arguments which have an underscore prefix... It's there to signify it's private. This has been put as a temporary measure after a discussion I had with @zdevito, but we never intended this functionality to be publicized and to have a lot of people depend on it...

Comment (dzhulgakov) [reply docs/source/notes/large_scale_deployments.rst]: Sorry, I missed this comment. I think passing metadata through PT files might be useful for higher level workflows built on top of PT. How do you feel about dropping underscore all together and making it official functionality? I don't see too many drawbacks as the interface name->blob is pretty generic and our container is a zip file already.

Of course, one can argue that additional files can be appended with just `ZipFile` python API. While it's ok for reading, there might be unintended effects for writing as ZipFile doesn't guarantee data block alignment (something our writer does). For reading, suggesting to use ZipFile should be sufficient.

**Pull request closed without merging**